contracts/BSCValidatorSet.sol

```diff
@@ -341,63 +372,54 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
-      }
-      uint n = currentValidatorSet.length;
-      for (uint i=index+1;i<n;i++) {
-        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
-      }
+
+    Validator memory validatorInfo = currentValidatorSet[index];
+    address validator = validatorInfo.consensusAddress;
+
+    if (
+      validator == address(0)                                       // - 0. check if empty validator
+      || (maxNumOfMaintaining == 0 || maintainSlashScale == 0)      // - 1. check if not start
+      || maintainingValidatorSet.length >= maxNumOfMaintaining      // - 2. check if exceeded upper limit
+      || validatorInfo.jailed                                       // - 3. check if jailed
+      || maintainInfoMap[validator].isMaintaining                   // - 4. check if maintaining
+      || maintainInfoMap[validator].startBlockNumber > 0            // - 5. check if has Maintained
+      || getValidators().length <= 1                                // - 6. check num of remaining unjailed currentValidators
+    ) {
+      return false;
     }
-    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    return true;
   }
 
-  function felony(address validator)external onlySlash override{
-    uint256 index = currentValidatorSetMap[validator];
-    if (index <= 0) {
-      return;
-    }
+  function enterMaintenance() external {
+    uint256 index = currentValidatorSetMap[msg.sender];
+    require(index > 0, "only current validators can enter maintenance");
+
     // the actually index
     index = index - 1;
-    uint256 income = currentValidatorSet[index].incoming;
-    uint256 rest = currentValidatorSet.length - 1;
-    if (rest==0) {
-      // will not remove the validator if it is the only one validator.
-      currentValidatorSet[index].incoming = 0;
-      return;
-    }
-    emit validatorFelony(validator,income);
-    delete currentValidatorSetMap[validator];
-    // It is ok that the validatorSet is not in order.
-    if (index != currentValidatorSet.length-1) {
-      currentValidatorSet[index] = currentValidatorSet[currentValidatorSet.length-1];
-      currentValidatorSetMap[currentValidatorSet[index].consensusAddress] = index + 1;
-    }
-    currentValidatorSet.pop();
-    uint256 averageDistribute = income/rest;
-    if (averageDistribute!=0) {
-      uint n = currentValidatorSet.length;
-      for (uint i=0;i<n;i++) {
-        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
-      }
-    }
-    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    require(canEnterMaintenance(index), "can not enter Temporary Maintenance");
+    _enterMaintenance(msg.sender, index);
+  }
+
+  function exitMaintenance() external {
+    require(maintainInfoMap[msg.sender].isMaintaining, "not in maintenance");
```

Comment:
We need more strong verification here, check whether it exist in `maintainingValidatorSet`

Reply:
fixed